### ISSUE TITLE
get frame render time when vsync is enabled

### ISSUE THREAD

**Issue by thmasn** [godotengine/godot-proposals#8399]
### Describe the project you are working on

i am working on a vr project, where maintaining the frame rate is crucial.
to provide the best possible visual quality, i want to scale the quality based on performance.

### Describe the problem or limitation you are having in your project

to make scaling based on game performance possible, the render time needs to be measurable.
as far as i can see, there is currently no way to get a read on this duration when vsync is enabled.

in a super fast scene, all measurement return a time delta which includes the wait time for vsync:
- using `delta` in `_process` gives the "clamped" vsync time of exactly 1/60s.
- using `Performance.get_monitor(Performance.TIME_PROCESS)` varies more, but seems to max out around 1/63s
- there seems to be no `beforeRender` or `afterRender` callback

there are already 2 issues for engine-level quality scaling, but simpler and more versatile GDscript based scaling should be easier to implement:
https://github.com/godotengine/godot-proposals/issues/917
https://github.com/godotengine/godot-proposals/issues/2183

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

- an additional monitor like `TIME_PROCESS_WITHOUT_VSYNCH` could be added.
- additional Methods `_before_render`, and `_after_render` could be added to the node class

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

- using `Performance.get_monitor(Performance.TIME_PROCESS_WITHOUT_VSYNCH)` would return the actual time a frame took to process physics and rendering, without the vsync wait time included

- using `_after_render()` in a Node would allow to perform actions after the frame has been rendered, and before the new frame starts. this would allow to measure the time manually

both would allow to write game-specific quality adjustments in GDscripts, so the target framerate can be achieved on all devices.

### can it be worked around with a few lines of script?

currently, this is not exposed to scripts, so cannot be solved here.
the only workaround is disabling vsync, which is not wanted in many cases

### Is there a reason why this should be core and not an add-on in the asset library?

this is a basic functionality present in most game engines

**Comment by thmasn** [comment]
the main issue is if we always want to maintain the vsync time like in vr:
we want to know when we are rendering faster than vsync, so we can increase the quality without dropping below the vsync rate.
currently, we can only get a read "we are faster than vsync", but "faster" could mean 1% faster (=>we do not want to increase quality), or 10x faster (=> we want to increase quality), so it is difficult to make a decision in the code if we should increase the quality.
all we can do is increase by p.e. 10%, and hope we do not drop down to 55fps, which results in using really small increments.

**Comment by Calinou** [comment]
> to make scaling based on game performance possible, the render time needs to be measurable.

This can already be measured as evidenced by the **View Frame Time** panel in the 3D editor viewport's **Perspective** menu :slightly_smiling_face: 

The methods to do this are:

```gdscript
var viewport_rid := get_viewport().get_viewport_rid()
var frametime_cpu := RenderingServer.viewport_get_measured_render_time_cpu(viewport_rid) + RenderingServer.get_frame_setup_time_cpu()
var viewport_gpu := get_viewport().get_viewport_rid().viewport_get_measured_render_time_gpu(viewport_rid)
```

See https://github.com/godot-extended-libraries/godot-debug-menu for a project that uses these methods.

Note: Do not change `Viewport.scaling_3d_scale` every frame, as this requires recreating the 3D buffers, which is slow and will lead to uneven frametimes.

**Comment by KoBeWi** [comment]
> there seems to be no beforeRender or afterRender callback

`RenderingServer.frame_pre_draw` and `RenderingServer.frame_post_draw`

**Comment by thmasn** [comment]
> > to make scaling based on game performance possible, the render time needs to be measurable.
> 
> This is already implemented since Godot 4.0. This can already be measured as evidenced by the **View Frame Time** panel in the 3D editor viewport's **Perspective** menu 🙂
> 
> The methods to do this are:
> 
> ```gdscript
> var viewport_rid := get_viewport().get_viewport_rid()
> 
> # Enable measurements (call once in `_ready()`).
> RenderingServer.viewport_set_measure_render_time(viewport_rid, true)
> 
> # Get measurements (call once every frame).
> var frametime_cpu := RenderingServer.viewport_get_measured_render_time_cpu(viewport_rid) + RenderingServer.get_frame_setup_time_cpu()
> var frametime_gpu := RenderingServer.viewport_get_measured_render_time_gpu(viewport_rid)
> ```
> 
> See https://github.com/godot-extended-libraries/godot-debug-menu for a project that uses these methods.
> 
> Do not change `Viewport.scaling_3d_scale` every frame, as this requires recreating the 3D buffers, which is slow and will lead to uneven frametimes. Likewise, be careful about changing some of the other graphics settings every frame, as they also require reallocating buffers on every change.

that is exactly what i needed. sorry for not finding it myself.
